chain/subscriber.go

```diff
@@ -0,0 +1,451 @@
+	}
+
+	ContractStoreSubscriber interface {
+		AddContractID(fcid types.FileContractID)
+	}
+
+	Subscriber struct {
+		cm     ChainManager
+		cs     ChainStore
+		logger *zap.SugaredLogger
+
+		announcementMaxAge time.Duration
+		walletAddress      types.Address
+
+		syncInterval    time.Duration
+		syncSig         chan struct{}
+		csUnsubscribeFn func()
+
+		mu             sync.Mutex
+		closedChan     chan struct{}
+		syncTicker     *time.Ticker
+		knownContracts map[types.FileContractID]api.ContractState
+	}
+)
+
+func NewSubscriber(cm ChainManager, cs ChainStore, contracts ContractStore, walletAddress types.Address, announcementMaxAge, syncInterval time.Duration, logger *zap.SugaredLogger) (_ *Subscriber, err error) {
+	if announcementMaxAge == 0 {
+		return nil, errors.New("announcementMaxAge must be non-zero")
+	}
+	if syncInterval == 0 {
+		return nil, errors.New("syncInterval must be non-zero")
+	}
+
+	// create chain subscriber
+	subscriber := &Subscriber{
+		cm:     cm,
+		cs:     cs,
+		logger: logger,
+
+		announcementMaxAge: announcementMaxAge,
+		walletAddress:      walletAddress,
+
+		syncInterval: syncInterval,
+		syncSig:      make(chan struct{}, 1),
+
+		closedChan: make(chan struct{}),
+	}
+
+	// subscribe ourselves to receive new contract ids
+	subscriber.knownContracts, subscriber.csUnsubscribeFn, err = contracts.AddContractStoreSubscriber(context.Background(), subscriber)
+	if err != nil {
+		return nil, err
+	}
+
+	return subscriber, nil
+}
+
+func (cs *Subscriber) AddContractID(id types.FileContractID) {
+	cs.mu.Lock()
+	defer cs.mu.Unlock()
+	cs.knownContracts[id] = api.ContractStatePending
+}
+
+func (cs *Subscriber) Close() error {
+	cs.mu.Lock()
+	defer cs.mu.Unlock()
+
+	// signal we are closing
+	close(cs.closedChan)
+
+	// unsubscribe from chain manager
+	cs.csUnsubscribeFn()
+
+	// stop ticker
+	if cs.syncTicker != nil {
+		cs.syncTicker.Stop()
+		select {
+		case <-cs.syncTicker.C:
+		default:
+		}
+	}
+
+	return nil
+}
+
+func (cs *Subscriber) Run() (func(), error) {
+	// perform an initial sync
+	index, err := cs.cs.ChainIndex()
+	if err != nil {
+		return nil, err
+	}
+	if err := cs.sync(index); err != nil {
+		return nil, fmt.Errorf("failed to subscribe to chain manager: %w", err)
+	}
+
+	// start trigger loop
+	cs.syncTicker = time.NewTicker(cs.syncInterval)
+	go func() {
+		for {
+			select {
+			case <-cs.closedChan:
+				return
+			case <-cs.syncTicker.C:
+			}
+			cs.triggerSync()
+		}
+	}()
+
+	// start sync loop
+	go func() {
+		for {
+			select {
+			case <-cs.closedChan:
+				return
+			case <-cs.syncSig:
+			}
+
+			ci, err := cs.cs.ChainIndex()
+			if err != nil {
+				cs.logger.Errorf("failed to get chain index: %v", err)
+				continue
+			}
+			err = cs.sync(ci)
+			if err != nil {
+				cs.logger.Errorf("failed to sync: %v", err)
+			}
+		}
+	}()
+
+	// trigger initial sync
+	cs.triggerSync()
+
+	// trigger a sync on reorgs
+	return cs.cm.OnReorg(func(_ types.ChainIndex) { cs.triggerSync() }), nil
```

Comment:
I don't think we need that actually. If we sync every X seconds anyway we will automatically handle a reorg.
That also makes it slightly safer since `triggerSync` is only ever called from a single thread.

Reply:
I don't see a reason to drop it though. I think instead we should poll every 10 minutes instead of every couple of seconds. Isn't that better than polling every second? I think it's kind of odd we design a subscription API that notifies us of reorgs and then just ignore it and resort to polling

Reply:
Alright just found out that `OnReorg` is actually not just happening on reorg so let's keep the trigger 😅